src/backend/distributed/utils/causal_clock.c

```diff
@@ -0,0 +1,759 @@
+	Oid outputFunctionId = FunctionOid("pg_catalog", "record_out", 1);
+	fmgr_info(outputFunctionId, outputFunction);
+	char *transactionId = OutputFunctionCall(outputFunction, transactionIdHeapDatum);
+
+	/* Pick the highest logical clock value among all transaction-nodes */
+	uint64 transactionClockValue = GetHighestClockInTransaction(transactionNodeList);
+
+	/* Persist the transactionId along with the logical commit-clock timestamp */
+	LogTransactionCommitClock(transactionId, transactionClockValue);
+
+	/* Adjust all the nodes with the new clock value */
+	AdjustClocksToTransactionHighest(transactionNodeList, transactionClockValue);
+}
+
+
+/*
+ * LogTransactionCommitClock registers the committed transaction along
+ * with the commit clock.
+ */
+static void
+LogTransactionCommitClock(char *transactionId, uint64 transactionClockValue)
+{
+	Datum values[Natts_pg_dist_commit_transaction];
+	bool isNulls[Natts_pg_dist_commit_transaction];
+	uint64 clockLogical = GET_LOGICAL(transactionClockValue);
+	uint32 clockCounter = GET_COUNTER(transactionClockValue);
+
+	ereport(DEBUG1, (errmsg("Persiting transaction %s with "
+							"clock logical (%lu) and counter(%u)",
+							transactionId, clockLogical, clockCounter)));
+
+	/* form new transaction tuple */
+	memset(values, 0, sizeof(values));
+	memset(isNulls, false, sizeof(isNulls));
+
+	uint64 transactionTimestamp = GetEpochTimeMs();
+
+	values[Anum_pg_dist_commit_transaction_transactionid - 1] =
+		CStringGetTextDatum(transactionId);
+	values[Anum_pg_dist_commit_transaction_clock_logical - 1] =
+		Int64GetDatum(clockLogical);
+	values[Anum_pg_dist_commit_transaction_clock_counter - 1] =
+		Int64GetDatum(clockCounter);
+	values[Anum_pg_dist_commit_transaction_timestamp - 1] =
+		Int64GetDatum(transactionTimestamp);
+
+	/* open transaction relation and insert new tuple */
+	Relation pgDistCommitTransaction =
+		table_open(DistCommitTransactionRelationId(), RowExclusiveLock);
+	TupleDesc tupleDescriptor = RelationGetDescr(pgDistCommitTransaction);
+	HeapTuple heapTuple = heap_form_tuple(tupleDescriptor, values, isNulls);
+
+	CatalogTupleInsert(pgDistCommitTransaction, heapTuple);
+
+	CommandCounterIncrement();
+
+	/* close relation and invalidate previous cache entry */
+	table_close(pgDistCommitTransaction, NoLock);
+}
+
+
+/*
+ * Initialize the shared memory clock value to the highest clock
+ * persisted. This will protect from any clock drifts.
+ */
+void
+InitClockAtBoot(void)
+{
+	if (creating_extension)
+	{
+		/* No catalog present yet */
+		return;
+	}
+
+	SpinLockAcquire(&logicalClockShmem->clockMutex);
+
+	/* Avoid repeated initialization */
+	if (logicalClockShmem->clockInitialized == true)
+	{
+		SpinLockRelease(&logicalClockShmem->clockMutex);
+		return;
+	}
+
+	/*
+	 * Set the flag before executing a distributed query
+	 * (else it might trigger this routine recursively.)
+	 */
+	logicalClockShmem->clockInitialized = true;
```

Comment:
should we do so after `	logicalClockShmem->clusterClockValue = START_NEW_CLOCK(epochValue);` ?

Reply:
Actually there are some synchronization and reentrant issue with this code, I've re-adjusted it.